services/rfq/relayer/inventory/rebalance.go

```diff
@@ -27,183 +27,185 @@ type RebalanceManager interface {
-// Note that only the given chain/token pair is considered for rebalance (as the destination chain).
-//
-//nolint:cyclop,nilnil
-func getRebalance(span trace.Span, cfg relconfig.Config, tokens map[int]map[common.Address]*TokenMetadata, chainID int, token common.Address) (rebalance *RebalanceData, err error) {
-	// get rebalance method
-	methods, err := cfg.GetRebalanceMethods(chainID, token.Hex())
+// getRebalances gets the best rebalance action for each token.
+func getRebalances(ctx context.Context, cfg relconfig.Config, inv map[int]map[common.Address]*TokenMetadata) (rebalances map[string]*RebalanceData, err error) {
+	rebalances = map[string]*RebalanceData{}
+
+	rebalanceCandidates, err := getRebalanceCandidates(ctx, cfg, inv)
 	if err != nil {
-		return nil, fmt.Errorf("could not get rebalance method: %w", err)
-	}
-	if len(methods) == 0 {
-		return nil, nil
+		return nil, fmt.Errorf("could not get rebalance candidates: %w", err)
 	}
 
-	// get token metadata
-	var rebalanceTokenData *TokenMetadata
-	for address, tokenData := range tokens[chainID] {
-		if address == token {
-			rebalanceTokenData = tokenData
-			break
+	for tokenName, methodCandidates := range rebalanceCandidates {
+		methodCandidatesSlice := []RebalanceData{}
+		for _, candidate := range methodCandidates {
+			if candidate == nil {
+				continue
+			}
+			methodCandidatesSlice = append(methodCandidatesSlice, *candidate)
 		}
+		rebalances[tokenName] = getBestRebalance(methodCandidatesSlice)
 	}
 
-	// evaluate the origin and dest of the rebalance based on min/max token balances
-	originTokenData, destTokenData, method := getRebalanceMetadatas(cfg, tokens, rebalanceTokenData.Name, methods)
-	if originTokenData == nil {
-		if span != nil {
-			span.SetAttributes(attribute.Bool("no_rebalance_origin", true))
-		}
-		return nil, nil
-	}
-	if destTokenData == nil {
-		if span != nil {
-			span.SetAttributes(attribute.Bool("no_rebalance_dest", true))
-		}
-		return nil, nil
-	}
-	if method == relconfig.RebalanceMethodNone {
-		if span != nil {
-			span.SetAttributes(attribute.Bool("no_rebalance_method", true))
-		}
-		return nil, nil
-	}
+	return rebalances, nil
+}
 
-	// if the given chain is not the destination of the rebalance, no need to do anything
-	if destTokenData.ChainID != chainID {
-		if span != nil {
-			span.SetAttributes(attribute.Int("rebalance_dest", destTokenData.ChainID))
+// getRebalanceCandidates gets the best rebalance for each token and rebalance method supported by the config.
+func getRebalanceCandidates(ctx context.Context, cfg relconfig.Config, inv map[int]map[common.Address]*TokenMetadata) (rebalances map[string]map[relconfig.RebalanceMethod]*RebalanceData, err error) {
+	rebalances = map[string]map[relconfig.RebalanceMethod]*RebalanceData{}
+	for chainID, chainCfg := range cfg.Chains {
+		for tokenName, tokenCfg := range chainCfg.Tokens {
+			if len(tokenCfg.RebalanceMethods) == 0 {
+				continue
+			}
```

Comment:
Do we need this check? And if yes, should we check `methods` length below instead to avoid relying on `TokenConfig` structure?

Reply:
Only reason I do this check first is to avoid inserting an empty map value into `rebalances` in the case that there are no valid rebalances for a given token. Not critical but I think slightly cleaner to only deal with non-empty maps